solver/llbsolver/solver.go

```diff
@@ -193,138 +197,186 @@ func (s *Solver) Solve(ctx context.Context, id string, sessionID string, req fro
 	}
 
 	var exporterResponse map[string]string
+	cacheExporters := exp.CacheExporters
 	if e := exp.Exporter; e != nil {
-		cached, err := result.ConvertResult(res, func(res solver.ResultProxy) (solver.CachedResult, error) {
-			return res.Result(ctx)
-		})
+		var inlineCacheExporter *RemoteCacheExporter
+		cacheExporters, inlineCacheExporter = splitCacheExporters(cacheExporters)
+		exporterResponse, err = exportWithInlineCacheExporter(ctx, e, inlineCacheExporter, j, res)
 		if err != nil {
 			return nil, err
 		}
+	}
 
-		inp, err := result.ConvertResult(cached, func(res solver.CachedResult) (cache.ImmutableRef, error) {
-			workerRef, ok := res.Sys().(*worker.WorkerRef)
-			if !ok {
-				return nil, errors.Errorf("invalid reference: %T", res.Sys())
-			}
-			return workerRef.ImmutableRef, nil
-		})
-		if err != nil {
-			return nil, err
+	cacheExporterResponse, err := runCacheExporters(ctx, cacheExporters, j, res)
+	if err != nil {
+		return nil, err
+	}
+
+	if exporterResponse == nil {
+		exporterResponse = make(map[string]string)
+	}
+
+	for k, v := range res.Metadata {
+		if strings.HasPrefix(k, "frontend.") {
+			exporterResponse[k] = string(v)
 		}
-		if res.Refs != nil {
-			m := make(map[string]cache.ImmutableRef, len(res.Refs))
-			for k, res := range res.Refs {
-				if res == nil {
-					m[k] = nil
-				} else {
-					r, err := res.Result(ctx)
-					if err != nil {
-						return nil, err
-					}
-					workerRef, ok := r.Sys().(*worker.WorkerRef)
-					if !ok {
-						return nil, errors.Errorf("invalid reference: %T", r.Sys())
-					}
-					m[k] = workerRef.ImmutableRef
-				}
-			}
-			inp.Refs = m
+		if strings.HasPrefix(k, exptypes.ExporterBuildInfo) {
+			exporterResponse[k] = base64.StdEncoding.EncodeToString(v)
+		}
+	}
+	for k, v := range cacheExporterResponse {
+		if strings.HasPrefix(k, "cache.") {
+			exporterResponse[k] = v
 		}
+	}
 
-		if _, ok := asInlineCache(exp.CacheExporter); ok {
-			if err := inBuilderContext(ctx, j, "preparing layers for inline cache", j.SessionID+"-cache-inline", func(ctx context.Context, _ session.Group) error {
-				if cached.Ref != nil {
-					dtic, err := inlineCache(ctx, exp.CacheExporter, cached.Ref, e.Config().Compression, session.NewGroup(sessionID))
-					if err != nil {
+	return &client.SolveResponse{
+		ExporterResponse: exporterResponse,
+	}, nil
+}
+
+func runCacheExporters(ctx context.Context, exporters []RemoteCacheExporter, j *solver.Job, res *frontend.Result) (map[string]string, error) {
+	eg, ctx := errgroup.WithContext(ctx)
+	g := session.NewGroup(j.SessionID)
+	var cacheExporterResponse map[string]string
+	resps := make([]map[string]string, len(exporters))
+	for i, exp := range exporters {
+		func(exp RemoteCacheExporter, i int) {
+			eg.Go(func() (err error) {
+				id := fmt.Sprint(j.SessionID, "-cache-", exp.Exporter.Name())
```

Comment:
I think we can use the index `i` instead of the exporter name here to distinguish between the different cache exporters.
```suggestion
				id := fmt.Sprint(j.SessionID, "-cache-", i)
```